src/raftstore/store/peer.rs

```diff
@@ -2007,44 +2008,74 @@ impl RequestInspector for Peer {
 }
 
 #[derive(Debug)]
-pub struct ReadExecutor<'r, 'e, 't> {
-    region: &'r metapb::Region,
-    engine: &'e Arc<DB>,
-    tag: &'t str,
+pub struct ReadExecutor {
+    check_epoch: bool,
+    engine: Arc<DB>,
+    snapshot: Option<SyncSnapshot>,
+    snapshot_time: Option<Timespec>,
 }
 
-impl<'r, 'e, 't> ReadExecutor<'r, 'e, 't> {
-    pub fn new(region: &'r metapb::Region, engine: &'e Arc<DB>, tag: &'t str) -> Self {
+impl ReadExecutor {
+    pub fn new(engine: Arc<DB>, check_epoch: bool) -> Self {
         ReadExecutor {
-            region,
+            check_epoch,
             engine,
-            tag,
+            snapshot: None,
+            snapshot_time: None,
+        }
+    }
+
+    #[inline]
+    pub fn snapshot_time(&mut self) -> Option<Timespec> {
```

Comment:
this function name seems a Getter function, but it needs mutable, strange.

Reply:
Because it takes snapshot lazily. It is used in single thread, I think it's fine to take mutable ref.